### DISCUSSION TITLE
Improvements to nesting navigators

### DISCUSSION THREAD

**Discussion by gantonious** [react-navigation/react-navigation#10491] [Ideas]
Hey everyone. I work on a fairly large app and have run into a few cases where I think nesting navigators could provide an elegant solution but doesn't work due to constraints made clear in the library [documentation](https://reactnavigation.org/docs/nesting-navigators/#best-practices-when-nesting). I opened this discussion to dive into these use cases and discuss potential solutions for them.

### Sharing screens across navigators

The app I work on is fairly interconnected and shares screens across our tabs. When I first came across wanting to share screens I naturally came up with the following pattern, grouping the shared screens into their own navigator and nesting it into the navigators I want to launch the screens from:

```jsx
const Tab = createBottomTabNavigator();

// An app with two tabs each with their own stack navigator
// that would like to share the screens in the Checkout navigator.
export default function App() {
  return (
    <NavigationContainer>
      <Tab.Navigator>
        <Tab.Screen name="Home" component={Home} />
        <Tab.Screen name="Explore" component={Explore} />
      </Tab.Navigator>
    </NavigationContainer>
  );
}

const HomeStack = createNativeStackNavigator();

function Home() {
  return (
    <HomeStack.Navigator>
      <HomeStack.Screen name="Feed" component={HomeFeedScreen} />
      <HomeStack.Screen name="Checkout" component={Checkout} />
    </HomeStack.Navigator>
  );
}

const ExploreStack = createNativeStackNavigator();

function Explore() {
  return (
    <ExploreStack.Navigator>
      <ExploreStack.Screen name="ExploreFeed" component={ExploreFeedScreen} />
      <ExploreStack.Screen name="Checkout" component={Checkout} />
    </ExploreStack.Navigator>
  );
}

const CheckoutStack = createNativeStackNavigator();

function Checkout() {
  return (
    <CheckoutStack.Navigator>
      <CheckoutStack.Screen name="CartSummary" component={CartSummaryScreen} />
      <CheckoutStack.Screen
        name="PaymentDetails"
        component={PaymentDetailsScreen}
      />
      <CheckoutStack.Screen
        name="PaymentConfirmation"
        component={PaymentConfirmationScreen}
      />
    </CheckoutStack.Navigator>
  );
}
```

Of course as mentioned above, this type of nesting (nesting like navigators) isn't really supported and is discouraged in the library documentation.

We have worked around this limitation by defining a list of all screens alongside their default options we want to share in a central location. Then in each navigator we want to access them in, we iterate over all the defined screens and add them as screens to the navigator. The solution works but doesn't really feel like React Navigation.

### Wrapping a group of screens in a context provider

Another interesting use case for nesting navigators I've come across is scoping a context provider to a specific group of screens. Say for this checkout navigator I want to share some state across all the entire checkout flow and only across the checkout flow. Ideally I could do something like this to achieve that:

```jsx
const CheckoutContext = createContext({...});

function Checkout() {
  return (
    <CheckoutContext.Provider value={/* Some shared state */}>
      <CheckoutStack.Navigator>
        <CheckoutContext.Screen name="CartSummary" component={CartSummaryScreen} />
        <CheckoutContext.Screen name="PaymentDetails" component={PaymentDetailsScreen} />
        <CheckoutContext.Screen
            name="PaymentConfirmation"
            component={PaymentConfirmationScreen}
        />
      </CheckoutStack.Navigator>
    </CheckoutContext.Provider>
  );
}
```

To summarize I know today React Navigation suggests against nesting navigators as a way to better organize your code but this constraint leads to:

- Increased difficulty sharing screens across multiple navigators.
- The inability to wrap a group of screens in a context provider.

### Idea

Without knowing too much about the technical constraints and the internals of React Navigation, would it be possible to detect the nesting of like navigators and delegate to the parent navigator to display the screens of the nested navigator? Some kind of navigator merging mechanism. That way we could have the parent navigator still handle all the UX and avoid any duplicate UI while still getting the code organization benefits nesting navigators can provide.

**Comment by satya164** [comment]
> scoping a context provider to a specific group of screens

I don't really see any benefit to this. It's not any different from having some data in a global store and components accessing the data they use. Regardless it's impossible to implement this in a good way without a custom API for context which is worse than using just React APIs.

> would it be possible to detect the nesting of like navigators and delegate to the parent navigator to display the screens of the nested navigator

Data flow goes from parent to child in React. A parent component doesn't know what's in a child component, at least not on the first render and not without callbacks. It makes it impossible or fragile to do this.

**Reply by gantonious** [reply]
> I don't really see any benefit to this. It's not any different from having some data in a global store and components accessing the data they use.

I think I gave a poor example for showcasing why we want this. For a checkout flow it would probably make sense to store your cart data globally.

The app I work on has several flows where we load data from the server for just that flow, edit it across multiple screens (hence the desire for a context), submit the changes, then discard that data. Being able to wrap that group of screens in a context provider would allow us just to store that state for the lifecycle of that flow. Being able to nest like navigators would provide us a good spot to wrap a group of screens with the appropriate context provider.

> Data flow goes from parent to child in React. A parent component doesn't know what's in a child component, at least not on the first render and not without callbacks. It makes it impossible or fragile to do this.

Yeah that makes sense. I was more thinking when the nested navigator is mounted we could check if we have a parent navigator of the same type and ask it to render our screens instead.
